lean4-infoview/src/infoview/rpcInterface.ts

```diff
@@ -183,3 +184,45 @@ export async function getGoToLocation(rs: RpcSessions, pos: DocumentPosition, ki
     const args: GetGoToLocationParams = { kind, info };
     return rs.call<LocationLink[]>(pos, 'Lean.Widget.getGoToLocation', args)
 }
+
+/** Represents an instance of a user widget that can be rendered.
+ * This is used as the input to the `UserWidget` component.
+ */
+export interface UserWidget {
+    /** The identifier of the type of widget to dynamically load. */
+    widgetSourceId: string
+    /** A hash (provided by Lean) of the widgetSource's sourcetext.
+     * This is used to look up the WidgetSource object.
+     */
+    hash : string
+    /** JSON object to be passed as props to the component */
+    props : any
+    range?: Range
+}
+
+/** The response type for the RPC call `Widget_getWidgets`. */
+export interface UserWidgets {
+    widgets : UserWidget[]
+}
+
+/** Given a position, returns all of the user-widgets on the infotree at this position. */
+export function Widget_getWidgets(rs: RpcSessions, pos: DocumentPosition): Promise<UserWidgets | undefined> {
+    return rs.call(pos, 'Lean.Widget.getWidgets', DocumentPosition.toTdpp(pos))
+}
+
+/** Code that should be dynamically loaded by the UserWidget component. */
+export interface WidgetSource {
+    /** JavaScript sourcecode. Should be a plain JavaScript ESModule whose default export is
+     * the component to render.
+     */
+    sourcetext: string
+    hash: string
+}
+
+/** Gets the static code for a given widget.
```

Comment:
```suggestion
/** Gets the static javascript code for a given widget.
```

@Vtec234 At some point you said that you called this function `getWidgetSourceJS` to clarify that it returns javascript, and so that we could potentially have `getWidgetSourceNeovim` in the future.  Did you have a change of heart?

Reply:
Originally it was called getWidgetStaticJS, but one conclusion of the RFC was that it should also work with other languages in the future, so the mention of JS should be removed from API. There is no check on the Lean side as to what language the source is. I guess we can add that, perhaps a language tag on source items?

Reply:
See also [here](https://github.com/leanprover/lean4/pull/1238#discussion_r907679198).

Reply:
Ok so the way this works now is that `UserWidget` has a field `javascript: String`, we can add more strings later if we really want. Then all of these source code strings are stored separately by their hash, and `getWidgetSource` just takes a hash and returns a string, but the hash is now stored as `javascriptHash`. Hence the getWidgetSource RPC just serves strings based on hash without knowing what type the string has.